File: JE/1_22/boj_9012.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class boj_9012 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int t = Integer.parseInt(br.readLine());

        while (t-- > 0) {
            String str = br.readLine();
            Stack<String> s = new Stack<>();

            boolean finish = false;
            for (int i = 0; i < str.length(); i++) {
                int input = str.charAt(i);
                if (input == '(')
                    s.push("(");
                else if (input == ')') {
                    if (s.empty()) {
                        finish = true;
                        break;
                    } else {
                        s.pop();
                    }
                }

            }

            if (!s.empty() || finish)
                System.out.println("NO");
            else
                System.out.println("YES");
        }
    }

}

File: Lim-YongKwan/1월 26일/BOJ_9012.cpp
#include<iostream>
#include<stack>
#include<string>
#define endl "\n"

using namespace std;

int main() {

	int n;
	cin >> n;

	while (n--) {
		stack<char> bracket;

		string s;
		cin >> s;
		bracket.push(s[0]);

		for (int i = 1; i < s.length(); i++) {
			if (bracket.size() != 0 && s[i] != bracket.top() && s[i] == ')') {
				bracket.pop();
			}
			else {
				bracket.push(s[i]);
			}
		}

		if (bracket.size() != 0) {
			cout << "NO" << endl;
		}
		else {
			cout << "YES" << endl;
		}
		bracket.empty();

	}
	return 0;
}